File: src/main/java/ru/urstannightmare/cathelpserver/configuration/ControllerAdvice.java
package ru.urstannightmare.cathelpserver.configuration;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.dao.DataRetrievalFailureException;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;
import ru.urstannightmare.cathelpserver.task.TaskNotFoundException;
import ru.urstannightmare.cathelpserver.task.dto.ErrorResponse;

@RestControllerAdvice
public class ControllerAdvice {
    private static final Logger log = LoggerFactory.getLogger(ControllerAdvice.class);

    @ExceptionHandler(value = {
            DataRetrievalFailureException.class,
            HttpMessageNotReadableException.class
    })
    public ResponseEntity<ErrorResponse> handleInvalidRequestArgumentException(
            Exception ex,
            WebRequest request) {

        log.error("Handled invalid request. {}", ex.getMessage());

        return new ResponseEntity<>(
                new ErrorResponse(HttpStatus.BAD_REQUEST.value(), ex.getMessage()),
                HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(
            value = TaskNotFoundException.class
    )
    public ResponseEntity<ErrorResponse> dataRetrievalFailureException(
            TaskNotFoundException ex,
            WebRequest  request
    ){
        return new ResponseEntity<>(
                new ErrorResponse(HttpStatus.NOT_FOUND.value(), ex.getMessage()),
                HttpStatus.NOT_FOUND
        );
    }


    @ExceptionHandler(
            value = MethodArgumentNotValidException.class
    )
    public ResponseEntity<ErrorResponse> handleInvalidArgumentException(
            Exception ex,
            WebRequest request) {
        var e = ((MethodArgumentNotValidException) ex);

        var errors = e.getDetailMessageArguments();
        StringBuilder builder = new StringBuilder();

        for (Object error : errors) {
            builder.append(error).append(" ");
        }

        log.error("Handled validation error. {}", ex.getMessage());

        return new ResponseEntity<>(
                new ErrorResponse(HttpStatus.BAD_REQUEST.value(), builder.toString()),
                HttpStatus.BAD_REQUEST
        );
    }

    @ExceptionHandler(
            value = DuplicateKeyException.class
    )
    public ResponseEntity<ErrorResponse> handleDuplicateKeyException(
            Exception ex,
            WebRequest request) {

        log.error("Handled duplicated key. {}", ex.getMessage());

        return new ResponseEntity<>(
                new ErrorResponse(HttpStatus.BAD_REQUEST.value(), "Key must be unique!"),
                HttpStatus.BAD_REQUEST
        );
    }
}

File: src/main/kotlin/ru/urstannightmare/cathelpserver/configuration/ControllerAdvice.kt
package ru.urstannightmare.cathelpserver.configuration

import jakarta.validation.ConstraintViolationException
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.WebRequest
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
import ru.urstannightmare.cathelpserver.configuration.security.AuthException
import ru.urstannightmare.cathelpserver.task.TaskNotFoundException
import ru.urstannightmare.cathelpserver.task.v2.dto.ErrorResponse
import java.util.stream.Collectors

@RestControllerAdvice
class ControllerAdvice {
    private val log = LoggerFactory.getLogger(ControllerAdvice::class.java)
    @ExceptionHandler(value = [TaskNotFoundException::class])
    fun dataRetrievalFailureException(
        ex: TaskNotFoundException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        return ResponseEntity(
            ErrorResponse(
                HttpStatus.NOT_FOUND.value(),
                ex.message ?: "No message",
                request.getDescription(false)
            ),
            HttpStatus.NOT_FOUND
        )
    }

    @ExceptionHandler(value = [MissingServletRequestParameterException::class])
    fun handleMussingParameterException(
        ex: MissingServletRequestParameterException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        log.warn("Handled ${request.getDescription(false)} request with missing ${ex.parameterName} parameter.")

        return ResponseEntity(
            ErrorResponse(
                HttpStatus.BAD_REQUEST.value(),
                "Missing [${ex.parameterName}] request parameter",
                request.getDescription(false)
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    @ExceptionHandler(
        value = [
            MethodArgumentTypeMismatchException::class,
            HttpMessageNotReadableException::class,
            ConstraintViolationException::class
        ]
    )
    fun handleIncorrectRequestParameterException(
        ex: RuntimeException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {

        var parameterName: Any? = when (ex) {
            is MethodArgumentTypeMismatchException -> ex.value
            is HttpMessageNotReadableException -> "requestBody"
            is ConstraintViolationException -> ex.constraintViolations
                .stream()
                .map { v -> v.propertyPath.toString()}
                .collect(Collectors.joining(","))

            else -> null
        }

        log.warn("Handled ${request.getDescription(false)} request with incorrect [$parameterName] parameter.")

        return ResponseEntity(
            ErrorResponse(
                HttpStatus.BAD_REQUEST.value(),
                "Incorrect request value [$parameterName]",
                request.getDescription(false)
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    @ExceptionHandler(
        value = [
            AuthException::class
        ]
    )
    fun handleAuthException(
        ex: RuntimeException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse>{
        log.warn("Handled bad auth attempt ${request.getDescription(false)}. {}", ex.message)

        return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
            .body(ErrorResponse(HttpStatus.UNAUTHORIZED.value(), "Auth error. ${ex.message}", ""))
    }
}